listenbrainz/webserver/static/js/src/stats/UserListeningActivity.tsx

```diff
@@ -0,0 +1,455 @@
+import * as React from "react";
+import MediaQuery from "react-responsive";
+
+import APIService from "../APIService";
+import Card from "../components/Card";
+import BarDualTone from "./BarDualTone";
+import Loader from "../components/Loader";
+
+export type UserListeningActivityProps = {
+  range: UserStatsAPIRange;
+  user: ListenBrainzUser;
+  apiUrl: string;
+};
+
+export type UserListeningActivityState = {
+  data: UserListeningActivityData;
+  lastRangePeriod: {
+    start?: Date;
+    end?: Date;
+  };
+  thisRangePeriod: {
+    start?: Date;
+    end?: Date;
+  };
+  loading: boolean;
+  totalListens: number;
+  avgListens: number;
+};
+
+export default class UserListeningActivity extends React.Component<
+  UserListeningActivityProps,
+  UserListeningActivityState
+> {
+  APIService: APIService;
+
+  rangeMap = {
+    week: {
+      dateFormat: {
+        weekday: "short",
+        timeZone: "UTC",
+      },
+      perRange: "day",
+    },
+    month: {
+      dateFormat: {
+        day: "2-digit",
+        timeZone: "UTC",
+      },
+      perRange: "day",
+    },
+    year: {
+      dateFormat: {
+        month: "short",
+        timeZone: "UTC",
+      },
+      perRange: "month",
+    },
+    all_time: {
+      dateFormat: {
+        year: "numeric",
+        timeZone: "UTC",
+      },
+      perRange: "year",
+    },
+  };
+
+  constructor(props: UserListeningActivityProps) {
+    super(props);
+
+    this.APIService = new APIService(
+      props.apiUrl || `${window.location.origin}/1`
+    );
+
+    this.state = {
+      data: [],
+      lastRangePeriod: {},
+      thisRangePeriod: {},
+      totalListens: 0,
+      avgListens: 0,
+      loading: false,
+    };
+  }
+
+  componentDidUpdate(prevProps: UserListeningActivityProps) {
+    const { range: prevRange } = prevProps;
+    const { range: currRange } = this.props;
+    if (prevRange !== currRange) {
+      this.loadData();
+    }
+  }
+
+  getData = async (): Promise<UserListeningActivityResponse> => {
+    const { range, user } = this.props;
+    try {
+      const data = await this.APIService.getUserListeningActivity(
+        user.name,
+        range
+      );
+      return data;
+    } catch (error) {
+      if (error.response && error.response.status === 204) {
+        const toThrow = new Error(
+          "Statistics for the user haven't been calculated yet."
+        );
+
+        this.setState(() => {
+          throw toThrow;
+        });
+      } else {
+        this.setState(() => {
+          throw error;
+        });
+      }
+    }
+    return {} as UserListeningActivityResponse;
+  };
+
+  processData = (
+    data: UserListeningActivityResponse
+  ): UserListeningActivityData => {
+    const { range } = this.props;
```

Comment:
If we get this range from the url, then we should validate it in this function. What happens if we pass "blah" as a range?

Reply:
Added a check for it.
https://drive.google.com/file/d/1-3ixUsvk9QzanE6qj0wTmOVBUxBIoOGv/view?usp=sharing